staging/src/k8s.io/apiserver/pkg/admission/attributes.go

```diff
@@ -32,19 +37,26 @@ type attributesRecord struct {
 	object      runtime.Object
 	oldObject   runtime.Object
 	userInfo    user.Info
+
+	annotations map[string]string
+	// other elements are always accessed in single goroutine.
+	// But ValidatingAdmissionWebhook sets annotations concurrently.
+	annotationLock sync.Mutex
 }
 
-func NewAttributesRecord(object runtime.Object, oldObject runtime.Object, kind schema.GroupVersionKind, namespace, name string, resource schema.GroupVersionResource, subresource string, operation Operation, userInfo user.Info) Attributes {
+func NewAttributesRecord(object runtime.Object, oldObject runtime.Object, kind schema.GroupVersionKind, namespace, name string, resource schema.GroupVersionResource, subresource string, operation Operation, userInfo user.Info, annotations map[string]string) Attributes {
```

Comment:
since the annotations object is just an accumulator, can we avoid changing the signature and just always initialize it to a nil or empty map?

Reply:
We need to pass annotation from one attribute to another:

https://github.com/kubernetes/kubernetes/blob/af58729c866abc6ebd08edf9e4c1e9bb5bbdc418/staging/src/k8s.io/apiserver/pkg/registry/rest/update.go#L241

I once introduced a new Method like: NewAttributesRecordWithAnnotation(), but it seems not that good.